### PR TITLE
Fix TransformedDistribution shaping logic

### LINKED ISSUE #21596: TransformedDistribution and event_shape
## 🐛 Bug

TransformedDistribution does not seem to have correct event_shape when the transform has changed the number of dimensions (e.g. with `StickBreakingTransform`).

## To Reproduce

```
import torch as tt
dist = tt.distributions.Dirichlet(tt.ones(3))
support = dist.support
tform = tt.distributions.constraint_registry.biject_to(support)
dist_unconstrained = tt.distributions.TransformedDistribution(dist,tform.inv)

print(tform)
print(dist_unconstrained.sample())
print(dist_unconstrained.event_shape)
```

Produces the results of

```
StickBreakingTransform()
tensor([ 0.3628, -1.9573])
torch.Size([3])
```

## Expected behavior

The last print statement should (seemingly) give:

```
torch.Size([2])
```

## Environment

 - PyTorch Version (e.g., 1.0): 1.1.0
 - OS (e.g., Linux): Mac
 - How you installed PyTorch (`conda`, `pip`, source): `conda`
 - Build command you used (if compiling from source): n/a
 - Python version: 3.6.4
 - CUDA/cuDNN version: n/a
 - GPU models and configuration: n/a
 - Any other relevant information: n/a

### LINKED ISSUE #34859: Incorrect shape from torch.distributions.kl.kl_divergence
## To Reproduce

Run the following code
```python
import torch
import torch.distributions as td

scale=torch.ones(2,3)
loc=torch.zeros(2,3)
normal = td.Normal(loc=loc, scale=scale)
diag_normal = td.Independent(normal, reinterpreted_batch_ndims=1)
trans_dist = td.TransformedDistribution(diag_normal, transforms=td.AffineTransform(loc=0., scale=2.))
print(td.kl.kl_divergence(trans_dist, trans_dist), "Got incorrect shape")
print(td.kl.kl_divergence(diag_normal, diag_normal), "Correct shape")
```

## Expected behavior

In the above code, the expected shape from the first kl_divergence() should be `[2]`. Instead, I got `[]`.

## Environment

ollecting environment information...
PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.2 LTS
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
CMake version: version 3.10.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.1.168
GPU models and configuration:
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti

Nvidia driver version: 440.33.01
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip3] numpy==1.18.1
[pip3] torch==1.4.0
[pip3] torchvision==0.5.0
[conda] Could not collect

##

It seems to be the bug is caused by the following code. It seems to me the commented line is correct.
https://github.com/pytorch/pytorch/blob/471ddacd8bea2f134e8d9d84e787c1612d9573a2/torch/distributions/kl.py#L438-L439

cc @vincentqb @fritzo @neerajprad @alicanb @vishwakftw

### LINKED ISSUE #50496: distributions.Independent does not correctly update distribution's support
## 🐛 Bug

When a batch dimension is reinterpreted to be an event dimension, the number of event dimensions should increase by 1 (e.g. a distribution over scalars becomes a distribution over vectors) and as a result the distribution's support should be updated. 

Because the distribution's support is not updated, transforms inferred by `biject_to` are applied elementwise (rather than summing/squeezing out the `event_dim`) and methods like `log_abs_det_jacobian` do not correctly compute the determinant of what now should be a multivariate transform (here, this multivariate log determinant is the sum of the elementwise log determinants).

Maybe we should call it `log_abs_diag_jacobian` here instead...? Just kidding.

## To Reproduce

When I run
```
import torch.distributions as dist

dist.Independent(dist.Normal(torch.zeros(1), torch.ones(1)), 1).support == dist.Normal(torch.zeros(1), torch.ones(1)).support
```
I find that the supports are identical before and after reinterpreting batch dimensions.

As a result, the following snippet:
```
import torch.distributions as dist
from torch.distributions.constraint_registry import biject_to

broken_dist = dist.Independent(dist.Exponential(torch.ones(2)), 1)
tform = biject_to(broken_dist.support)

x = torch.randn(3,2)
y = tform(x)  # NOTE: this is applied element-wise across x
tform.log_abs_det_jacobian(x, y).shape. # because broken_dist.event_shape == [2], after taking determinant this should be [3]
```
should report `torch.Size([3])` because taking the log determinant removes the `event_dim`, but instead it reports `[3,2]`.

## Expected behavior

After reinterpreting batch dimensions, the left-hand side is now a distribution over 1-dimensional rank-one tensors. Therefore, I expect the support to be `constraints.real_vector` in this example. More generally, I expect the support of a `dist.Independent` to be a product over the reinterpreted batch dimension of the individual supports, and for the inferred `transforms` to properly sum over the new `event_dim`.

## Environment

```
Collecting environment information...
PyTorch version: 1.7.0
Is debug build: True
CUDA used to build PyTorch: 11.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.19.1

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.4
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] botorch==0.3.3
[pip3] gpytorch==1.3.0
[pip3] numpy==1.19.4
[pip3] numpyro==0.4.1
[pip3] torch==1.7.0
[pip3] torchvision==0.8.1
[conda] Could not collect
```

## Additional context

cc @fritzo @neerajprad @alicanb @vishwakftw @nikitaved

### PR DESCRIPTION
**Pull request by fritzo**
Fixes #50496
Fixes #34859
Fixes #21596

This fixes many bugs involving `TransformedDistribution` and `ComposeTransform` when the component transforms changed their event shapes. Part of the fix is to introduce an `IndependentTransform` analogous to `distributions.Independent` and `constraints.independent`, and to introduce methods `Transform.forward_shape()` and `.inverse_shape()`. I have followed @fehiepsi's suggestion and replaced `.input_event_dim` -> `.domain.event_dim` and `.output_event_dim` -> `.codomain.event_dim`. This allows us to deprecate `.event_dim` as an attribute.

## Summary of changes

- Fixes `TransformDistribution` and `ComposeTransform` shape errors.
- Fixes a behavior bug in `LogisticNormal`.
- Fixes `kl_divergence(TransformedDistribution, TransformedDistribution)`
- Adds methods `Transform.forward_shape()`, `.inverse_shape()` which are required for correct shape computations in `TransformedDistribution` and `ComposeTransform`.
- Adds an `IndependentTransform`.
- Adds a `ReshapeTransform` which is invaluable in testing shape logic in `ComposeTransform` and `TransformedDistribution` and which will be used by @stefanwebb flowtorch.
- Fixes incorrect default values in `constraints.dependent.event_dim`.
- Documents the `.event_dim` and `.is_discrete` attributes.

## Changes planned for follow-up PRs

- Memoize `@constraints.dependent_property` as we do with `@lazy_property`, since we now consult those properties much more often.

## Tested
- [x] added a test for `Dist.support` vs `Dist(**params).support` to ensure static and dynamic attributes agree.
- [x] refactoring is covered by existing tests
- [x] add test cases for `ReshapedTransform`
- [x] add a test for `TransformedDistribution` on a wide grid of input shapes
- [x] added a regression test for #34859

cc @fehiepsi @feynmanliang @stefanwebb

### CODE REVIEW

File: torch/distributions/transforms.py
Comment (fritzo) [inline torch/distributions/transforms.py:269]: This section is tricky.

File: torch/distributions/transforms.py
Comment (fritzo) [inline torch/distributions/transforms.py:284]: This section is tricky.

File: torch/distributions/transforms.py
Comment (fehiepsi) [reply torch/distributions/transforms.py:269]: Yeah, agree. I am proving that this is equivalent to [here](https://github.com/pyro-ppl/numpyro/pull/876/files#diff-be93e2377765c460352859d72dc154f1bedb2791dda750b5bea634e518e834d4R169) but it is not clear for now. :))

File: torch/distributions/transformed_distribution.py
Comment (fehiepsi) [inline torch/distributions/transformed_distribution.py:79]: This should belong to the computation of `event_dim` in the constructor.

File: torch/distributions/transforms.py
Comment (fritzo) [reply torch/distributions/transforms.py:269]: @fehiepsi here's an example to show why shifting is required. Consider a fictional transform
```py
t = SquareMatrixToVectorTransform()
assert t.domain.event_dim == 2
assert t.codomain.event_dim == 1
```
Now consider
```py
c = ComposeTransform([t, t, t, t, DiscreteCosineTransform(), t.inv, t.inv, t.inv, t.inv])
assert c.domain.event_dim == 4
assert c.codomain.event_dim == 4
```
What `c` does is to flatten a 4-tensor into a vector, mix up all the elements, then reshape it.

Comment (fehiepsi) [reply torch/distributions/transforms.py:269]: Yes, I agree that we need it. I just meant that
```python
            event_dim += part.domain.event_dim - part.codomain.event_dim
            event_dim = max(event_dim, part.domain.event_dim)
```
is equivalent to
```python
            event_dim = part.domain.event_dim + max(event_dim - part.codomain.event_dim, 0)
```

File: torch/distributions/transformed_distribution.py
Comment (fritzo) [reply torch/distributions/transformed_distribution.py:79]: Good point! Actually the constructor needs to do a full shape computation, not merely a dim computation. Thus I believe we will need to implement `Transform.forward_shapes()`. I'll add that...

Comment (fehiepsi) [reply torch/distributions/transformed_distribution.py:79]: Agreed!

File: torch/distributions/transforms.py
Comment (fehiepsi) [reply torch/distributions/transforms.py:269]: By the way, I think you will need to adjust log_abs_det_jacobian (see the [current implementation](https://github.com/pyro-ppl/numpyro/pull/876/files#diff-be93e2377765c460352859d72dc154f1bedb2791dda750b5bea634e518e834d4R220-R234) in NumPyro) here and in TransformedDistribution.

File: torch/distributions/transforms.py
Comment (fehiepsi) [inline torch/distributions/transforms.py:185]: I am not sure why we need to forward both `batch_shape` and `event_shape`? Shouldn't `forward_shapes(shape)` enough?

File: torch/distributions/transformed_distribution.py
Comment (fritzo) [reply torch/distributions/transformed_distribution.py]: Done.

File: torch/distributions/transforms.py
Comment (fritzo) [reply torch/distributions/transforms.py:188]: Well we need helpers to compute one of the following representations of metadata:
either `(batch_shape, event_shape)` or `(shape, event_dim)`. I have opted for `(batch_shape, event_shape)` since that fits in a single function. That representation also makes the `TransformedDistribution` computations nice. I suppose we could change `.forward_shape()` to compute directly on `shape`, but then in most places we use it we would need to manually examine `.domain.event_dim` or `.codomain.event_dim`. 🤷

File: torch/distributions/transforms.py
Comment (fritzo) [reply torch/distributions/transforms.py:306]: Done.

File: torch/distributions/transforms.py
Comment (fritzo) [reply torch/distributions/transforms.py:188]: Let me code up your suggested `.forward_shape(shape)` version to see if it's any simpler...

Comment (fritzo) [reply torch/distributions/transforms.py]: @fehiepsi you're right, `.forward_shape(shape)` is much simpler. I just pushed. Thanks 😄

File: torch/distributions/transforms.py
Comment (fritzo) [inline torch/distributions/transforms.py:436]: Note @stefanwebb requests that `in_shape` be lazy if possible, to reduce possibility of errors by users, and to be closer to the syntax of `Tensor.reshape(out_shape)`

**Comment by neerajprad** [comment]
Could you merge in changes from the earlier PR?

File: torch/distributions/transforms.py
Comment (fehiepsi) [inline torch/distributions/transforms.py:718]: Nice trick to deal with scalars!!!

File: torch/distributions/transforms.py
Comment (neerajprad) [inline torch/distributions/transforms.py:418]: Should we validate that the return value has at least `reinterpreted_batch_ndims` dims (otherwise, it will only error when we pass it to `log_abs_det_jacobian`)?

Comment (fritzo) [reply torch/distributions/transforms.py:420]: Done

File: torch/distributions/transforms.py
Comment (fritzo) [reply torch/distributions/transforms.py:449]: Actually `ReshapeTransform` with only output shape would not be a valid transform since the inverse transform would would not be defined. I think a better solution would be for flowtorch to define a lazy `Reshape(out_shape)` that only lazily constructs a `ReshapeTransform(in_shape, out_shape)` once the `in_shape` is known.

**Comment by neerajprad** [comment]
The changes to the `IndependentTransform` look great. I'll defer to @stefanwebb and @fehiepsi for the ReshapeTransform related changes. I'm importing this internally to verify that this doesn't break anything on the client side, and will do another closer review later in the evening.

File: torch/distributions/transformed_distribution.py
Comment (fehiepsi) [inline torch/distributions/transformed_distribution.py:71]: pretty clean!

**Review by fehiepsi — DISMISSED** [review]

**Review by fehiepsi — COMMENTED** [review]
@fritzo You missed the `forward_shape` and `inverse_shape` for ExponentTransform and SoftmaxTransform. The former needs broadcasting to include `exponent` shape. The latter should be like StickBreakingTransform. Other changes look great to me!

**Review by neerajprad — APPROVED** [review]
Thanks for adding detailed tests! LGTM, pending comments from @fehipsi's.

**Comment by fritzo** [comment]
@fehiepsi @neerajprad thanks for detailed review!

> You missed the `forward_shape` and `inverse_shape` for `ExponentTransform` and `SoftmaxTransform`

I don't think PyTorch has an `ExponentTransform`, that must be NumPyro only 😄 . I've added a check to `SoftmaxTransform.forward_shape()` and `.inverse_shape()` but note that transform preserves shape exactly, unlike `StickBreakingTransform` which changes the size of the rightmost axis.

**Comment by fehiepsi** [comment]
Oops... I meant `PowerTransform` @fritzo...

**Review by fehiepsi — APPROVED** [review]

**Pull request closed without merging**
